scripts/detection/matrix_net/train_matrix_net.py

```diff
@@ -0,0 +1,315 @@
+    if args.num_samples < 0:
+        args.num_samples = len(train_dataset)
+    return train_dataset, val_dataset, val_metric
+
+def get_dataloader(net, train_dataset, val_dataset, data_shape, batch_size, num_workers, ctx):
+    """Get dataloader."""
+    width, height = data_shape, data_shape
+    num_class = len(train_dataset.classes)
+    # 96 = 1(image) + 19 (number of matrix layers) * 5 (heatmap, wh_target, wh_mask, offset_target, offset_mask)
+    batchify_fn = Tuple([Stack() for _ in range(96)])  
+    train_loader = gluon.data.DataLoader(
+        train_dataset.transform(MatrixNetDefaultTrainTransform(
+            width, height, num_class=num_class, layers_range=net.layers_range)),
+        batch_size, True, batchify_fn=batchify_fn, last_batch='rollover', num_workers=num_workers)
+    val_batchify_fn = Tuple(Stack(), Pad(pad_val=-1))
+    val_loader = gluon.data.DataLoader(
+        val_dataset.transform(MatrixNetDefaultValTransform(width, height)),
+        batch_size, False, batchify_fn=val_batchify_fn, last_batch='keep', num_workers=num_workers)
+    return train_loader, val_loader
+
+def save_params(net, best_map, current_map, epoch, save_interval, prefix):
+    current_map = float(current_map)
+    if current_map > best_map[0]:
+        best_map[0] = current_map
+        net.save_parameters('{:s}_best.params'.format(prefix, epoch, current_map))
+        with open(prefix+'_best_map.log', 'a') as f:
+            f.write('{:04d}:\t{:.4f}\n'.format(epoch, current_map))
+    if save_interval and epoch % save_interval == 0:
+        net.save_parameters('{:s}_{:04d}_{:.4f}.params'.format(prefix, epoch, current_map))
+
+def validate(net, val_data, ctx, eval_metric, flip_test=False):
+    """Test on validation dataset."""
+    eval_metric.reset()
+    net.flip_test = flip_test
+    mx.nd.waitall()
+    net.hybridize()
+    for batch in val_data:
+        data = gluon.utils.split_and_load(batch[0], ctx_list=ctx, batch_axis=0, even_split=False)
+        label = gluon.utils.split_and_load(batch[1], ctx_list=ctx, batch_axis=0, even_split=False)
+        det_bboxes = []
+        det_ids = []
+        det_scores = []
+        gt_bboxes = []
+        gt_ids = []
+        gt_difficults = []
+        for x, y in zip(data, label):
+            # get prediction results
+            ids, scores, bboxes = net(x)
+            det_ids.append(ids)
+            det_scores.append(scores)
+            # clip to image size
+            det_bboxes.append(bboxes.clip(0, batch[0].shape[2]))
+            # split ground truths
+            gt_ids.append(y.slice_axis(axis=-1, begin=4, end=5))
+            gt_bboxes.append(y.slice_axis(axis=-1, begin=0, end=4))
+            gt_difficults.append(y.slice_axis(axis=-1, begin=5, end=6) if y.shape[-1] > 5 else None)
+
+        # update metric
+        eval_metric.update(det_bboxes, det_ids, det_scores, gt_bboxes, gt_ids, gt_difficults)
+    return eval_metric.get()
```

Comment:
I didn't save the errror message. I remember the error message started roughly from these lines and was related to data type. The error message was something like 'float64 is not supported in pycocotools'. I was confused about the error because I didn't think I use the data type mentioned. Maybe can run the code to reproduce the bug.

Reply:
it seems to be an issue with latest numpy that breaks pycocotools. Maybe related: https://github.com/facebookresearch/detectron2/issues/580